internal/controllers/plans.go

```diff
@@ -73,70 +78,72 @@ func (s Server) GetPlanByID(ctx echo.Context) error {
-		return model.Error(ctx, "invalid plan description", http.StatusBadRequest)
-	}
-	var req = model.Plan{Name: plan.PlanName, Description: plan.Description}
-	err = s.GORMDB.Debug().Create(&req).Error
-	if err != nil {
-		return model.Error(ctx, err.Error(), http.StatusInternalServerError)
+	if err = plan.Validate(); err != nil {
+		return model.Error(ctx, err.Error(), http.StatusBadRequest)
 	}
-	return model.Success(ctx, "Success", http.StatusOK)
+	// Begin a transaction.
+	return s.GORMDB.Transaction(func(tx *gorm.DB) error {
+		dbPlan := plan.ToDBModel()
+		// Look up each resource type and update it in the struct.
+		for i, planQuotaDefault := range dbPlan.PlanQuotaDefaults {
+			resourceType, err := db.GetResourceTypeByName(tx, planQuotaDefault.ResourceType.Name)
+			if err != nil {
+				return model.Error(ctx, err.Error(), http.StatusInternalServerError)
+			}
+			if resourceType == nil {
+				msg := fmt.Sprintf("resource type not found: %s", resourceType.Name)
+				return model.Error(ctx, msg, http.StatusBadRequest)
+			}
+			dbPlan.PlanQuotaDefaults[i].ResourceType = *resourceType
+		}
+		// Add the plan to the database.
+		err := tx.Create(&dbPlan).Error
+		if err != nil {
+			return model.Error(ctx, err.Error(), http.StatusInternalServerError)
+		}
+		return model.SuccessMessage(ctx, "Success", http.StatusOK)
+	})
 }
 
 func (s Server) AddPlanQuotaDefault(ctx echo.Context) error {
-	planName := "Basic"
-	resourceName1 := "CPU"
-	cpuValue := 4.00
-	resourceName2 := "STORAGE"
-	storageValue := 1000.00
-	var plan = model.Plan{Name: planName}
-	err := s.GORMDB.Debug().Find(&plan, "name=?", planName).Error
-	if err != nil {
-		return ctx.JSON(http.StatusInternalServerError,
-			model.ErrorResponse("plan name not found", http.StatusInternalServerError))
+	var err error
+	// Parse and validate the request body.
+	var planQuotaDefaultValues PlanQuotaDefaultValues
+	if err = ctx.Bind(&planQuotaDefaultValues); err != nil {
+		return model.Error(ctx, err.Error(), http.StatusBadRequest)
 	}
-	planId := *plan.ID
-	var cpu = model.ResourceType{Name: resourceName1}
-	err = s.GORMDB.Debug().Find(&cpu, "name=?", resourceName1).Error
-	if err != nil {
-		return ctx.JSON(http.StatusInternalServerError,
-			model.ErrorResponse("resource Type not found: "+resourceName1, http.StatusInternalServerError))
+	if planQuotaDefaultValues.PlanName == "" {
+		return model.Error(ctx, "plan name can not be empty", http.StatusBadRequest)
+	}
+	if planQuotaDefaultValues.ResourceTypeName == "" {
+		return model.Error(ctx, "resource type name can not be empty", http.StatusBadRequest)
 	}
-	cpuId := *cpu.ID
-	var req = model.PlanQuotaDefault{PlanID: &planId, ResourceTypeID: &cpuId, QuotaValue: cpuValue}
-	err = s.GORMDB.Debug().Create(&req).Error
+
+	plan, err := db.GetPlan(s.GORMDB, planQuotaDefaultValues.PlanName)
 	if err != nil {
-		return model.Error(ctx, err.Error(), http.StatusInternalServerError)
+		return model.Error(ctx, "PLan", http.StatusBadRequest)
```

Comment:
There's a typo on this line, and it would be nice if the error message were a little more descriptive. `GetPlan` also returns `nil` for the plan if it doesn't exist. It would be helpful to check for this condition as well in order to prevent this endpoint from returning a confusing error message. The catch is that the check for a nil plan has to happen after the check for a non-nil error.